### ISSUE TITLE
New release

### ISSUE THREAD

**Issue by jmbowman** [pytest-dev/pytest-django#317]
The ``--keepdb`` fix from #261 looks like it'll save about 4-7 minutes on each test run of a project I'm working on (depending on hardware); is there a timeline yet for a new pytest-django release which would include it?  I don't see any issues currently tagged as release blockers.

**Comment by blueyed** [comment]
Does --reuse-db not work for you?

**Comment by jmbowman** [comment]
It does until a new migration is added; then you need to explicitly drop the ``--reuse-db`` parameter in order to apply the migration (making for a slow test run), and then add it back afterwards (so the next time is slow again because the last run didn't keep the database).  The advantage of ``--keepdb`` is that it runs any unapplied migrations before running the tests, so:

* You don't need to change the parameters used to run the tests depending on whether a new migration was added or not
* Applying just the new migrations once is much faster than 2 test runs applying them all
* You still need to recreate the DB if you change or remove a migration in development, but that's true of both flags

This drawback of ``--reuse-db`` in the current release is annoying but tolerable when running tests manually, but a royal pain when running continuous integration tests on git commits.  I have multiple Jenkins jobs set up to run tox like this for different packages, and there's no good way to add the flag just when the commit contains no new migrations (so the flag is always off and it's always slow).  I've been trying to cut down on the duration of those jobs to shorten the feedback loop on code changes, and this looks like a simple way to make some of them up to 50% faster.

Just in case it wasn't clear, I think this change was already made in the PR I mentioned (to have ``--reuse-db`` actually use ``--keepdb`` behind the scenes on Django 1.8+).  I'd just like it to be in an official release to avoid needing to package up a private release (and to aid people for whom that's an even bigger hassle because they haven't already set up a private package server).

**Comment by blueyed** [comment]
Thanks for your explanation.

Although I've added the change myself, I was not aware of it's benefits.. :)
I drop the test db in case of missing migrations, which makes --reuse-db recreate it.
FWIW, are you aware of the --nomigrations flag?

Apart from that a new release would be nice to have anyway.

**Comment by jmbowman** [comment]
``--nomigrations`` works well in many cases, but I have some projects which have tests to verify that database constraints created by SQL in migrations (because Django models don't yet have a way to define partial unique indexes, etc.) are working as intended and handled correctly when violated.  Those tests consistently fail if you skip the migrations altogether, since the constraint never gets created.

**Comment by blueyed** [comment]
> since the constraint never gets created

I see.  For this you could have pytest fixtures instead / additionally: the fixture would either use the data from the DB/migration or create it to be available for the tests.

**Comment by jmbowman** [comment]
Yes, and that's roughly what I did pre-1.7 when it was tougher to integrate migrations into the test cycle.  But that meant duplicating the code for making the relevant database changes, and introduced a risk that you're testing something which doesn't quite match what you're actually applying in the migrations for real installations.

**Comment by blueyed** [comment]
Yeah.  Thanks for elaborating again.

@pelme 
What do you think about having a new release? :)

**Comment by pelme** [comment]
I will try to put out a new release in the next couple of days. I would like to test this change out with my project properly (I use --reuse-db quite extensively).

Btw, I also have problems with slow CI builds because of migrations and would like to improve things further:

* I did a quick hack the other day calculating a md5 hash of all migration files which was fairly easy. The md5 hash could be saved in a special/hidden table in the database and then be used to re-run all migrations when any migration file actually changed. I think that would give 100 % accuracy on when to re-run migrations and speedy runs all other times.

* When using xdist: I would like to make it possible to create a single database from the migrations and then copy that database. Currently when I use i.e. `-n4`, all migrations runs in 4 processes which is a real performance killer. Doing the migrations in a single process and then creating 4 copies with `CREATE DATABASE ... TEMPLATE ...;` would probably speed things up a lot.

**Comment by blueyed** [comment]
For what it's worth, you could also use something like this to skip Travis builds for migration tests (instead of the md5 hash):

    if git diff --quiet --exit-code "$TRAVIS_COMMIT_RANGE" \
        project/app/migrations project/app/tests/test_migrations.py; then
            ...

**Comment by blueyed** [comment]
btw: https://github.com/fastmonkeys/stellar is really nice in this regard, too: it creates DB snapshots.
But there was some issue when I've tried to integrate it with pytest-django / our CI: https://github.com/fastmonkeys/stellar/issues/53.

**Comment by blueyed** [comment]
I've just tried using master (with the `--keepdb` patch), instead of dropping the test database, but that resulted in an error:

> psycopg2.ProgrammingError: relation "django_content_type" already exists

Maybe related to using `--nomigrations` also?

The hopefully relevant part of the traceback:

```
File "…/pyenv/project/lib/python3.5/site-packages/_pytest/python.py", line 2057, in call_fixture_func
  res = fixturefunc(**kwargs)
File "…/pytest-django/pytest_django/fixtures.py", line 57, in _django_db_setup
  interactive=False, **db_args)
File "…/django/django/test/runner.py", line 726, in setup_databases
  serialize=connection.settings_dict.get("TEST", {}).get("SERIALIZE", True),
File "…/django/django/db/backends/base/creation.py", line 70, in create_test_db
  run_syncdb=True,
File "…/django/django/core/management/__init__.py", line 119, in call_command
  return command.execute(*args, **defaults)
File "…/django/django/core/management/base.py", line 399, in execute
  output = self.handle(*args, **options)
File "…/django/django/core/management/commands/migrate.py", line 200, in handle
  executor.migrate(targets, plan, fake=fake, fake_initial=fake_initial)
File "…/django/django/db/migrations/executor.py", line 92, in migrate
  self._migrate_all_forwards(plan, full_plan, fake=fake, fake_initial=fake_initial)
File "…/django/django/db/migrations/executor.py", line 121, in _migrate_all_forwards
  state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
File "…/django/django/db/migrations/executor.py", line 198, in apply_migration
  state = migration.apply(state, schema_editor)
```

**Comment by blueyed** [comment]
Ok, for clarification: `--nomigrations` skips trying to apply the migration, but without it the above error happens.

**Comment by sasha0** [comment]
@blueyed did you run tests without `--nomigrations` and with `--keepdb` option on the fresh database?

**Comment by blueyed** [comment]
@sasha0 
Looks like it, but I don't remember - sorry for the late response.

Anyway, the issue itself is fixed: there's a 3.0 release now.. :tada:
